libsolidity/ast/Types.cpp

```diff
@@ -700,6 +700,42 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 				value = rational(denominator, numerator);
 			break;
 		}
+		case Token::SHL:
+		{
+			using boost::multiprecision::pow;
+			if (isFractional() || other.isFractional())
+				return TypePointer();
+			else if (other.m_value < 0)
+				return TypePointer();
+			else if (other.m_value > numeric_limits<uint32_t>::max())
+				return TypePointer();
+			uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
+			// shortcut
+			if (exponent > 256) {
```

Comment:
We should allow much larger exponents. For the DoS, I think it is better to estimate the size of the resulting value and throw an error if it is likely to take too much memory. Returning zero is not an option.